### PR TITLE
Use Xcode project for macOS builds

### LINKED ISSUE #523: Wrong NSBundle information for packaged app
**Describe the bug**
When packaging an app with briefcase on macOS, the information returned by `NSBundle` at runtime is incorrect. This seems to be because the actual executable is `Contents/Resources/Support/bin` while `Contents/MacOS/Hello World` is only a wrapper script.

This has significant implications for the capabilities of the resulting app. Notably, sending desktop notifications with either `NSUnserNotificationCenter` or `UNUsernotificationCenter` will not work because neither can correctly identify the app from which it is used. In fact, calling `UNUsernotificationCenter.currentNotificationCenter()` results in a segfault with an assertion error.

**To Reproduce**
Run through the beeware tutorial to create a sample app and modify app.py as follows:

```Python
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW

from toga_cocoa.libs import NSBundle


class HelloWorld(toga.App):

    def startup(self):

        l0 = toga.Label(f"Bundle ID: {NSBundle.mainBundle.bundleIdentifier}")
        l1 = toga.Label(f"Bundle path: {NSBundle.mainBundle.bundlePath}")
        l2 = toga.Label(f"Resource path: {NSBundle.mainBundle.resourcePath}")

        main_box = toga.Box(children=[l0, l1, l2], style=Pack(direction=COLUMN))

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()


def main():
    return HelloWorld()

```

This gives the following bundle info:

| Key             | Value                                                                                                     |
|-----------------|-----------------------------------------------------------------------------------------------------------|
| Bundle ID:      | None                                                                                                                                     |
| Bundle path:         | ~/beeware-tutorial/helloworld/macOS/Hello World/Hello World.app/Contents/Resources/Support/bin|
| Resource path:  | ~/beeware-tutorial/helloworld/macOS/Hello World/Hello World.app/Contents/Resources/Support/bin                     |

**Expected behavior**
This is expected:

| Key             | Value                                                                                                     |
|-----------------|-----------------------------------------------------------------------------------------------------------|
| Bundle ID:      | com.example.helloworld                                                                                    |
| Bundle path:         | ~/beeware-tutorial/helloworld/macOS/Hello World/Hello World.app |
| Resource path:  | ~/beeware-tutorial/helloworld/macOS/Hello World/Hello World.app/Contents/Resources  |

**Environment:**
 - Operating System: macOS 11
 - Python version: 3.8
 - Software versions:
   - Briefcase: 0.3.3
   - Toga: 0.3.0.dev25

**Additional context**
This seems to be quite a fundamental problem and I am unaware of a good way to solve it. It may be possible to patch NSBundle at runtime to return the correct values but this is not particularly elegant.

It seems like you are already somewhat aware of the problem, as this hackish workaround to get the resource path shows:

https://github.com/beeware/toga/blob/adac30ba4ce048e333cd5b0f2080582f906cad53/src/cocoa/toga_cocoa/app.py#L118

### PR DESCRIPTION
**Pull request by SamSchott**
This PR fixes #523 by building a native binary linked against libPython as the application binary instead of an executable shell script. This replicates the build process for iOS. In particular, this entails:

* A new cockiecutter template for a macOS Xcode project, available at https://github.com/SamSchott/briefcase-macOS-Xcode-template.
* The "app" output format on macOS has been replaced by a "xcode" output format.
* The macOS build of the Python-Apple-support package has been changed out generate a static library, similar to the iOS build. The changes are available at https://github.com/SamSchott/Python-Apple-support, a PR for this will follow.
* The macOS package command to create a dmg has been adapted to work with the new xcode format.

At the moment, the build follows closely what is done on iOS: a fully functional Xcode project is generated from the cockiecutter template and populated with the app code, dependencies and resources. This is then built, producing a native binary and the final app.

There are two reasons why I have decided to create a full Xcode project instead of providing the binary as part of the template:

1. The binary currently contains the name of the Python module to start. This is "hard-coded" and must therefore be set before the build. However, there may be ways around that.

2. I think it is actually very powerful to have an Xcode project template that works with Python programs. Briefcase currently is not very customisable and may never allow for certain platform-specific options. However, I can see a case where briefcase is used to create the Xcode project and the end user then can modify the project manually / programatically to link additional frameworks, add sandboxing, etc. 

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
First off - apologies for the slow review on this. I've been moving house, and packing + unpacking has consumed much of my time. 

Code wise, this looks great; can't fault it. I had a few headaches actually testing it (first working out which branch of the support package had the fixes, then dealing with issues caused by an Xcode upgrade, and then getting a build that didn't try to drag gettext into the library); but once I had those sorted out, it worked a treat.

One (relatively minor) note inline; I'm open to discussion on whether what I'm suggesting is a good or bad idea.

As for the larger design decision to include a project, rather than a hard-coded binary. At the very least, I agree that it makes sense to at least include it as an option, because the sort of customisability that affords is, as you've noted, considerable. 

The major benefit to *not* needing the project is that it removes the need to download and install Xcode to get an app running. Even though it's a trivial install from the App Store, you'd be surprised how much of an impediment it is for some people - especially for education uses. I've lost track of the number of people who have asked "do I *really* need to download Xcode" even for iOS deploys. If we *can* avoid it for macOS, I think it's worth it for the end-developer experience.

The fact that the module name needs to be hard-coded in the binary doesn't need to be a show stopper - it should be possible to address by either rewriting the string in the binary, or having the binary read the resources plist file (or some other metadata file) and extract the name of the executable module.

So: my inclination would be to include both "patched binary/app" and "project/xcode" versions as options for macOS. This does lead to the question of how to handle the DMG target - which makes me wonder if we've perhaps done the wrong thing with DMG handling. 

I originally set up dmg as a subclass of app. I'm wondering if it shouldn't be the other way around. This would make app and Xcode specialisations of dmg; dmg is then treated as a packaging option for any macOS app, with a `--no-dmg` or `--app-only` option available for people who want to distribute a raw app rather than a DMG. Thoughts? Ordinarily, I'd treat this as a separate problem to what you're doing here - except that this PR will remove the `app` target, which would make re-introducing it later somewhat difficult.

File: tests/platforms/macOS/dmg/test_package.py [hunk 1]
```diff
@@ -25,8 +25,8 @@ def test_build_dmg(first_app_config, tmp_path):
         [
             'codesign',
             '--sign', 'Sekrit identity (DEADBEEF)',
-            '--entitlements', str(tmp_path / 'macOS' / 'First App' / 'Entitlements.plist'),
-            '--deep', str(tmp_path / 'macOS' / 'First App' / 'First App.app'),
+            '--entitlements', str(command.entitlements_path(first_app_config)),
+            '--deep', str(command.binary_path(first_app_config)),
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply tests/platforms/macOS/dmg/test_package.py:29]: > it's effectively duplicated by the mixin tests

This was my reasoning for not spelling out the paths explicitly. When I changed the `binary_path` I encountered so many test failures that it was difficult to find out what I actually needed to fix / change. And changing it was a somewhat annoying job of copy-pasting.

This way, there will be one clear failure when changing the path while the actual commands are still running as intended...

**Comment by SamSchott** [comment]
No worries about the delay, I hope the move went well!

On both of your points: 

1. Yes, I agree that having both options of a full Xcode project and a pre-built app bundle would be nice. I'll see what I can do...

2. Regarding the order of subclassing the commands: I was a bit confused about the initial layout myself but decided to just keep it as is. My fundamental issue lies with treating App, Xcode and DMG as output formats on equal footing. To me, App and Xcode are both output formats (which already yield installable artefacts by themselves) while DMG is just a choice of packaging or maybe even publishing. I would not feel bad about relegating it to a mixin for the packaging or publishing of App and Xcode.

**Comment by SamSchott** [comment]
I take your point and agree that it's nice to have both options: a full Xcode project and a pre-complied app.

And I wasn't really sure what to do about the DMG target either. They way I see it:

**Comment by SamSchott** [comment]
Just closed this accidentally when trying to clear a comment :/

**Comment by SamSchott** [comment]
@freakboy3742, before I proceed to work on this, I've got two questions:

1. What would be the best way to create a cookiecutter template for pre-built app bundle? Just build it locally, hope to get everything right once, and push it to github? This would require some amount of manual work every time we want to update the binary. Or create a second repository the basic Xcode project and scripts to build it and convert it into a template?

2. What is your opinion on the dmg target? Keep it but change the inheritance order or relegate it to a command line option for the packaging steps of the Xcode and App targets?

**Comment by freakboy3742** [comment]
@SamSchott My first thought on project structure for the pre-built app would be a separate *branch* on the same repo (similar to how GitHub pages uses an orphan branch for source vs published versions of pages), with a CI process that builds and pushes to the template branches. If we're *really* clever, we might even be able to use the Xcode template to generate the binary for the App template...

As for DMG - I'm inclined to say it should be a formatting option on the packaging step. i.e. something like `briefcase package macOS Xcode --format=dmg` would become legal. My inclination is that `app` (prebuilt binary) with DMG packaging should be the default, with `--format=raw` (just the app) being an option.

**Comment by SamSchott** [comment]
This is slowly coming to together.

1. I have updated the briefcase-macOS-template at https://github.com/SamSchott/briefcase-macOS-template to automatically generate a pre-built app template from the Xcode template. The structure of that repo is as follows now:
   * master branch: contains only readme and other information
   * dev branch: contains cookiecutter xcode-template and a stub for the app-template, a script [scripts/xcode2app.py](https://github.com/SamSchott/briefcase-macOS-template/blob/dev/scripts/xcode2app.py) to generate a full app-template with a pre-built executable from the xcode template, and a GitHub action to push the result to a new branch.
   * 3.7-xcode, 3.7-app, etc: branches containing the final templates for different Python versions
2. I have resurrected the app target and relegated the dmg target to a packaging option, as discussed.

I'm still a bit ambivalent about the structure on two points: First, briefcase currently expects a different repo to provide the template for each target. For now, I have included a hackish workaround which allows manually specifying a repo + branch for the target (6a9f369) to enable testing. But keeping with the current structure, we'd want to push the final templates to separate repos. Second, the clear separation between template and support package now becomes blurred since the pre-compiled binary is part of the template. But I guess this is a necessary consequence of our approach.

They may be a couple of loose ends to tie up. Especially getting Xcode project right without overlooking anything is difficult. I tried to keep as close to the iOS-Xcode-template as possible but may have made some mistakes. Then there is the wealth of config files: `cookiecutter.json` and `briefcase.toml` files for both templates, `info.plist`, etc. It's easy to make mistakes there and overlook them in review. I don't envy you going through all of this to check!

Care to have another look at this, both for the general structure and the actual implementation of automatic template generation?

**Comment by SamSchott** [comment]
@freakboy3742, I know that reviewing such large sets of changes is a lot of work. But would you have time to have another look at this? It will be easier for me to make changes while the code is still fresh in my memory...

**Comment by freakboy3742** [comment]
@SamSchott Apologies for not reviewing this one faster - I'll put it on my todo list for this weekend.

File: src/briefcase/platforms/macOS/app.py [hunk 2]
```diff
@@ -25,10 +22,25 @@ def binary_path(self, app):
...
 class macOSAppCreateCommand(macOSAppMixin, CreateCommand):
     description = "Create and populate a macOS app."
 
+    def install_app_support_package(self, app: BaseConfig):
+        """
+        Install the application support packge.
+
+        :param app: The config object for the app
+        """
+        # nothing to do here template already contains built binary
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/briefcase/platforms/macOS/app.py:41]: I see you've put the standard library in the template; we will want to avoid that, as it couples the template to a specific Python version.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is looking pretty solid. I've flagged some fairly minor things inline, but the core organization of the code looks good to me. The only things that I can see that might needs some more work are things you've mostly already flagged as workaround/hacks, or relate to the templates themselves.

1. The handling of templates and branches. Personally, I don't see a problem with there being 2 different template repos - at the end of the day, they're two fundamentally different project structures, and it's almost an accident that the executable artefact produced by the Xcode template can be used in the App template.

2. The inclusion of the support package in the template content. It looks like you've included the support package content in the App template. Was there a particular reason for this? My 

3. The templates themselves. I found a couple of minor bugs in the templates (mostly icon handling, and that sort of thing), and some files that I'm not completely sure need to be in the template (some signing artefacts, a PkgInfo file, and an Assets.car file). These will probably be easier to sort out if we can get them in the form of a formal PR against a repo; if you're don't have any objection to the "2 template repository" approach, I can set up a near-empty briefcase-macOS-Xcode-template repo, and we can stage these the template updates to support this PR as PRs against their related template repos.

File: src/briefcase/commands/create.py [hunk 3]
```diff
@@ -253,15 +253,23 @@ def generate_app_template(self, app: BaseConfig):
...
+        parts = app.template.split("@")
+        if len(parts) > 1:
+            app.template = parts[0]
+            branch = parts[-1]
+        else:
+            branch = self.python_version_tag
+
+        print("Using app template: {app_template}, branch={branch}".format(
             app_template=app.template,
+            branch=branch,
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/create.py:265]: Hm, that's a tricky question. @saroad2's solution from #519 is more explicit and I therefore tend to prefer it. On the other handing, specifying the branch with "@" syntax is something that developers have become very used to.

I've also seen the discussion around forcing developers to name their branches following the Python version. Again, this is difficult. I personally prefer the flexibility of completely overriding the branch name. There are however caveats when using a template or support package for a Python version different from the one used to run briefcase. One issue can be the pip-install of `app_packages`, which will parse `python_version` flags in the dependencies according to the running Python version and not the support package.

[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/create.py:265]: On the second point, a stern warning ("here be dragons") in the documentation should be sufficient.

File: src/briefcase/commands/package.py [hunk 4]
```diff
@@ -48,6 +48,14 @@ def _package_app(self, app: BaseConfig, update: bool, **options):
         return state
 
     def add_options(self, parser):
+        parser.add_argument(
+            '-f',
+            '--format',
+            dest='format',
+            help='Output format to use (dmg or raw).',
+            default='dmg',
+            action='store_false',
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/package.py:57]: Good point.

File: src/briefcase/platforms/macOS/__init__.py [hunk 5]
```diff
@@ -19,3 +33,206 @@ def verify_tools(self):
         # Verify superclass tools *after* xcode. This ensures we get the
         # git check *after* the xcode check.
         super().verify_tools()
+
+
+class macOSPackageMixin:
+
+    def distribution_path(self, app):
+        return self.platform_path / '{app.formal_name}-{app.version}.dmg'.format(
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/platforms/macOS/__init__.py:41]: Ah, darn. Is there a good way to handle this? We won't really know the `distribution_path` until the user has packages the app.

File: src/briefcase/platforms/macOS/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/platforms/macOS/app.py:41]: >  we will want to avoid that, as it couples the template to a specific Python version.

Do we want to avoid that? The template is already firmly coupled to the Python version, considering that the binary contains a statically linked Python interpreter. Or would you rather move the binary into a custom "app" support package? 

This is probably one more reason to move to a framework build. I wish I had more experience with that but this is really above my pay grade...

**Comment by SamSchott** [comment]
@freakboy3742, thank you for the review!

Regarding your points:

I agree on the 2 template repo approach. There is indeed nothing firmly linking the Xcode and the App templates, apart from the "coincidence" that we can derive one from the other. If you create the repos I'll open PRs against them.

Regarding the inclusion of the Support Package in the template, my reasoning was as follows: The binary, which is part of the template, already includes the Python interpreter. I thought there is no harm in going all the way and including the standard lib as well. But I am not married to this approach. If you'd rather copy the lib during the build process, that's fine by me. It would be even better if we could fully decouple the template and the Python version. Maybe this is possible when using a framework build? I am not sure.

Agreed, it's easier to sort the bugs in the template when having separate PRs. I thought I had fixed all the icon template issues... Xcode seems to insist on signing the app (if only ad-hoc) when building a release version. If there is no way around that, the signature can always be stripped afterwards.

**Comment by freakboy3742** [comment]
In case you missed it - I created a stub [briefcase-macOS-Xcode-template](https://github.com/beeware/briefcase-macOS-Xcode-template) yesterday (and then forgot to mention that I'd done so). 

I'd prefer to keep the template and the libraries separate; that gives the end-user the flexibility to build their own support package if they want (for example, they could build a stripped down standard library, or include components that we don't ship by default, like Tkinter).

**Comment by SamSchott** [comment]
Thanks, I did miss that! Ok, a separate library it is then.

File: src/briefcase/commands/package.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/package.py:57]: Also, there is a name conflict with the 'show_output_formats' argument.

File: tests/platforms/macOS/xcode/test_select_identity.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
+from unittest import mock
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply tests/platforms/macOS/xcode/test_select_identity.py:1]: Yes, pretty much. Apart from the `DummyPublishCommand` inheritance.

File: src/briefcase/platforms/macOS/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/platforms/macOS/__init__.py]: I have changed `distribution_path` to take a `package_format` argument.

File: src/briefcase/commands/create.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/create.py]: Ok, I have rebased to include the changes from #519.

File: src/briefcase/commands/package.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/commands/package.py]: I've changed the argument name to `packge_format` and given the choices 'app' and 'dmg'.

File: tests/platforms/macOS/xcode/test_select_identity.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SamSchott) [reply tests/platforms/macOS/xcode/test_select_identity.py]: I have removed this duplicate test now.

**Comment by SamSchott** [comment]
I have now made the following changes:

1. Removed the Python lib from the app template.
2. Addressed your in-line comments.
3. Opened PRs against the `briefcase-macOS-Xcode-template` repo at https://github.com/beeware/briefcase-macOS-Xcode-template/pull/1 and the existing `briefcase-macOS-app-template` repo at https://github.com/beeware/briefcase-macOS-app-template/pull/4. I have included the script and GH action to generate the app template in the second PR since the those are specific to the app template itself.

Regarding icon handling in the template: it seems to work correctly for me. Maybe you can elaborate in the corresponding PR?

The stray files: I have removed the signing step from the build process. Building the Xcode template will now result in an unsigned app bundle. The Assets.car and PkgInfo files are automatically generated during the build process and part of every native macOS app. We can probably remove them without any side effects. We don't use AppKit's asset handling and can save 1 - 2 MB without the Assets.car file. Since this is an explicit deviation from the platform default, I have initially kept them in.

**Comment by SamSchott** [comment]
One more step in the right direction: I have opened a PR to include the static lib in the macOS support package at https://github.com/beeware/Python-Apple-support/pull/108.

**Comment by SamSchott** [comment]
I've fixed a few Python 3.5 compatibility issues, added tests for the macOS `install_app_support_package` command, and updated the documentation.

This should be ready for review now.

**Comment by SamSchott** [comment]
@freakboy3742, could you have another look at this? I've been using this briefcase branch with the modified support package in production for while now its been building quite smoothly.

**Comment by freakboy3742** [comment]
@SamSchott Apologies for dragging my feet on this - my last few weeks have been incredibly hectic. I promise I haven't forgotten about this patch; I'm just struggling to find the time I need to give it the attention it needs. With any luck, my schedule will clear up in a week or two.

**Review by freakboy3742 — APPROVED** [review]
Apologies for the long delay before I got to reviewing this - the patch is great. I've flagged a couple of minor typos (which I can clean up if you don't get to them); and one "maybe nice to have" if you're so inclined - but otherwise, I can't find fault with what you've got here. 

I need to do a bigger review of the app templates themselves, and the changes to the support package - in particular, the fact that we are now in a position to support M1 hardware means we may need to make some tweaks to the Xcode template (and possibly the invocation of xcodebuild?).

I'm hoping to get some time this weekend to dig into the fine details of those templates; with any luck, I'll be able to roll a new Briefcase release including these changes once that review is done.

**Comment by SamSchott** [comment]
Thanks for the positive review! And yes, the support packages will need to be tweaked to support universal binaries.

While using this branch in practice I have encountered another issue with the output directories: both the Xcode and the App target create directories `self.platform_path / app.formal_name`. This can be problematic if a user would like both targets to coexist and also causes issues when issuing a sequence of commands with different targets. For example:

```console
$ briefcase create macOS xcode
$ briefcase build macOS app
```

The second command will attempt to build the app from the Xcode template because it sees that `self.bundle_path(app)` already exists.

What do you think of appending the target type to the `bundle_path` for all platforms?

**Comment by freakboy3742** [comment]
That's a good point. Adding the output format as an additional `bundle_path` component is the obvious way to disambiguate; I'm not wild about the very long path that results... but I can't think of a better fix, either.

File: src/briefcase/platforms/macOS/__init__.py [hunk 7]
```diff
@@ -19,3 +33,202 @@ def verify_tools(self):
...
         # git check *after* the xcode check.
         super().verify_tools()
+
+
+class macOSPackageMixin:
+
+    def verify_tools(self):
+        super().verify_tools()
+        if dmgbuild is None:
+            raise BriefcaseCommandError("A macOS DMG can only be created on macOS.")
```
[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/platforms/macOS/__init__.py:43]: That's a valid point. The same would hold for the general `macOSMixin` which requires command line tools to be installed. When building with the "app" target, those are really only required for packaging / signing.

[the remark this answers is not in the document]
Comment (SamSchott) [reply src/briefcase/platforms/macOS/__init__.py:43]: Actually, this check should be overhauled more generally. Packing on macOS will do either of the following at the moment:

1. Sign the app (requires Xcode command line tools).
2. Optionally build a DMG (requires dmgbuild, always installed on macOS).

Both are only possible on macOS. So the macOS packaging command should *always* fail on different platforms and needs to check for command line tools as well.

**Comment by SamSchott** [comment]
I have now added the output format as a `bundle_path` component after the platform which is the its logical place in the hierarchy IMO. So now the path is "platform / format / app name".

I've also moved the `verify_tools` checks to the respective commands where they are actually needed. It is therefore now possible to "build" an app target on other platforms. Just the packaging / signing will raise errors.

Finally, the failing tests are due to a change in how setuptools reads the entrypoints from setup.cfg. This is fixed in #575.

**Comment by SamSchott** [comment]
Ok, I’m done with changes now. The last couple of commits appear a bit erratic because I’ve made using the GitHub website from my iPad. But their essence is:

1. Actually verify that Xcode is installed for the Xcode target (instead of just the command line tools).
2. Improve the `test_run_app_failed` tests by raising a `CalledProcessError` and checking that it is handled correctly instead of directly raising a `BriefcaseCommandError`.
3. Pass strings instead of path-like objects to `subprocess.run`. This enables compatibility with Python 3.7 and earlier on Windows.

Let me know if you are happy with all of this or if I have overlooked something.

**Comment by freakboy3742** [comment]
I've been looking at this over the weekend; here's a status update so you know where I'm at:

The Xcode project is working great; I've pushed versions of the Xcode template to the briefcase-macOS-Xcode-template repo, and I've built (but not published) support packages for macOS in the "static lib" format.

I hit a couple of minor bugs, stemming from the change including the project type in the bundle path. The iOS backend, for example, is manually constructing the path of the executable and not using the `bundle_path` shortcut; as a result, you can build, but not run an iOS project. I've pushed the fixes I've found so far.

The only sticking point at the moment is the app-based project. The core pieces work (although I needed to wrestle for a while to get icons right; and, for some reason, the plist file *must* have NSPrincipalClass defined, or else the app renders with fuzzy fonts...). I need to work out how to integrate the 'auto update binary' part of the process - more specifically what the *trigger* for updating the binary should be. I'm almost inclined to say that it should be tied to pushing a new support package, except that would tie the support package to Briefcase. 

I've also got a small backwards compatibility issue. I'm not overly concerned about backwards compatibility per se - users of a v0.3 product should expect a few roadbumps - but when I push the new support packages and templates, there will be a small window where anyone with an *old* version of briefcase will start getting some weird errors, because the layout of the support package is slightly different. I *think* this can be managed - I just need to get the possible scenarios straight in my head.

tl;dr - I think we're getting close, and we'll be in a much stronger place when this lands - I just need to sort the last few minor issues. Thanks so much for your effort on this patch - and apologies again for taking so long to get around to reviewing it.

**Comment by SamSchott** [comment]
> The iOS backend, for example, is manually constructing the path of the executable and not using the bundle_path shortcut; as a result, you can build, but not run an iOS project. I've pushed the fixes I've found so far.

Ah, yes, I somehow missed that. Thanks for spotting it!

> The only sticking point at the moment is the app-based project. The core pieces work (although I needed to wrestle for a while to get icons right; and, for some reason, the plist file must have NSPrincipalClass defined, or else the app renders with fuzzy fonts...).

The icon part is indeed a bit confusing because the Xcode and app templates currently expect different icon sets to be provided. I mostly followed the iOS Xcode template here. It would still be nice to unify the icon requirements for all macOS targets.

> I need to work out how to integrate the 'auto update binary' part of the process - more specifically what the trigger for updating the binary should be. I'm almost inclined to say that it should be tied to pushing a new support package, except that would tie the support package to Briefcase.

Tying the support package to the app template is inherent in the way that the app template works now, so this should be acceptable for the time being. In the longer term, distributing the support package as a framework build will decouple it from the app template again. So I wouldn't worry about it too much.

**Comment by freakboy3742** [comment]
Another update:

I've audited the other backends and made the path changes required.

I've worked out how to make the project layout historically backwards compatible. The paths inside the app are a little more convoluted (`Support/Python/Resources/lib`) - but this means that someone with an old version of briefcase, using the *new* version of the template will still end up with a working app. The app will definitely be bigger than it needs to be, but an upgrade will fix that. 

The dev branches of the app and Xcode templates now contain working (backwards compatible) templates for Python 3.9; the Xcode project and app binary have also been updated to generate universal2 (i.e., Intel + Apple Silicon) binaries. The CI config on the app config should be enough to work, but I can't fully test it until it's merged, and there's a new version of Briefcase to drive the build.

So - if you're able to replicate success with this branch, those templates, and a custom build of the support package (I've *tagged* a new release for the support packages, but haven't pushed them yet), I'll merge this and publish and new Briefcase release.

**Comment by SamSchott** [comment]
I've just tested both the Xcode and the app template. The Xcode template works well but the app template has PYTHONHOME  set to `Contents/Resources/Support/Python/Resources`. There is one too many "Resources" at the end of the path.

I also see that you have kept the shorter path structure for the Xcode template with the standard lib at `Contents/Resources/Python`. Having those different structures will make automatic conversion a bit more complicated.

**Comment by SamSchott** [comment]
Yes, I am using the dev branches for both builds. And the executable in the app template is definitely a fat binary.

Just to clarify, the directory structure of the app bundle is as expected. The issue is that `PYTHONPATH` and `sys.base_prefix` have the wrong path set when the interpreter is initialised.

It's a bit late here but I can investigate further tomorrow.

**Comment by SamSchott** [comment]
I just double checked and the init routine is definitely setting the wrong path. This is the startup output from running the executable directly:

```
Python path configuration:
  PYTHONHOME = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  PYTHONPATH = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app:/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app_packages'
  program name = 'python3'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/MacOS/Maestral'
  sys.base_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.base_exec_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.platlibdir = 'lib'
  sys.executable = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/MacOS/Maestral'
  sys.prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.exec_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.path = [
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app_packages',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python39.zip',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python3.9',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python3.9/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x0000000108dc1e00 (most recent call first):
<no Python frame>
```

I've also manually downloaded the executable from the dev branch and had a look at it. It does contain code to set the PYTHONHOME to `%@/Support/Python/Resources/. Is it possible that you uploaded a different version than the one which you have locally?

**Comment by SamSchott** [comment]
I just double checked and the init routine is definitely setting the wrong path. This is the startup output from running the executable directly:

```
Python path configuration:
  PYTHONHOME = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  PYTHONPATH = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app:/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app_packages'
  program name = 'python3'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/MacOS/Maestral'
  sys.base_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.base_exec_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.platlibdir = 'lib'
  sys.executable = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/MacOS/Maestral'
  sys.prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.exec_prefix = '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources'
  sys.path = [
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/app_packages',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python39.zip',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python3.9',
    '/Users/samschott/Python/maestral-cocoa/macOS/app/Maestral/Maestral.app/Contents/Resources/Support/Python/Resources/lib/python3.9/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x0000000108dc1e00 (most recent call first):
<no Python frame>
```

I've also manually downloaded the executable from the dev branch and had a look at it. It does contain code to set the PYTHONHOME to `%@/Support/Python/Resources/. Is it possible that you uploaded a different version than the one which you have locally?

**Comment by freakboy3742** [comment]
I'm confused - what are you saying is the right and wrong paths? Those logs all seem to be pointing at `Maestral.app/Contents/Resources/Support/Python/Resources/lib`, which AFAICT is the *correct* path, and is consistent with the support package currently being generated by the dev branch.

**Comment by SamSchott** [comment]
🤦Never mind me, it's been a few busy days here. The path *is* correct and the support package is also extracted correctly. It's just that I was using a Python 3.8 support package. This worked just fine with the Xcode template but messed up the app template. The issue wasn't with `sys.prefix` but rather with `sys.path` pointing at `/lib/python3.9`.

 I guess that the statically linked Python lib in the app template is version 3.9?

**Comment by freakboy3742** [comment]
Heh - no worries - we've all been there :-)

And yes - the binary in the app template dev branch is currently 3.9-based; once the dev template is merged into the version branches, CI *should* recompile and update the binaries for each version.

**Comment by SamSchott** [comment]
Ok, I can confirm that the app template is working as intended on my end when using the Python 3.9 lib :)
